### ISSUE TITLE
Stacked Borrows "is this argument" diagnostics go wrong for protected in-place function arguments

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/miri#3051]
For example, in [this test case](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6fa0b3d435bfc26dd3007b3ba316cf5c), or in [this one](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0cd23c8fcbfa5802baac3fc8e57be5db), it doesn't even point at something that is an argument.

Cc @saethlin

**Comment by saethlin** [comment]
🤨 Is the problem here just that custom MIR doesn't have coherent spans? Is it supposed to? My instinct is that anyone using custom MIR has already opted out of usable diagnostics.

**Comment by RalfJung** [comment]
No, the problem is that we're using protectors for things that are not function arguments. When a function argument is passed by-move, then we put a protector on the memory it comes from. We also put a protector on the return place. The diagnostics assume that only function arguments are protected and then things go downhill from there.

Reliably triggering these errors requires custom MIR, but with MIR optimizations it should be possible to also trigger those errors without nightly features.

**Comment by saethlin** [comment]
Ah I see! My brain grouped the diagnostic with the wrong span in the error. At first glance this seems pretty thorny to do a good diagnostic for. Hm.

**Comment by RalfJung** [comment]
Yeah it's definitely tricky, and also fairly low-priority since it cannot usually be encountered by users.
